### PR TITLE
[AST] Clean up handling of single-expression closures

### PR DESCRIPTION
**Pull request by DougGregor**
Stop treating single-expression closures as special throughout most of the compiler, with the exception of the constraint solver where the semantics *are* intentionally different. This eliminates a   bit of special-case code and paves the way toward experimental support for multi-statement closures checked within their enclosing expressions.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -3659,16 +3659,24 @@ static CallExpr *findTrailingClosureTarget(SourceManager &SM,
...
   Finder.resolve();
-  if (Finder.getContexts().empty()
-      || !Finder.getContexts().back().is<Expr*>())
+  auto contexts = Finder.getContexts();
+  if (contexts.empty())
     return nullptr;
-  CallExpr *CE = cast<CallExpr>(Finder.getContexts().back().get<Expr*>());
+
+  // If the innermost context is a brace statement, drop it.
+  if (contexts.back().is<Stmt *>()) {
```
Comment (AnthonyLatsis) [inline lib/IDE/Refactoring.cpp:3667]: Is the check for just `Stmt` intentional here, given the comment?

File: lib/Sema/ConstraintSystem.h [hunk 2]
```diff
@@ -5493,6 +5493,12 @@ BraceStmt *applyFunctionBuilderTransform(
 /// of the parameter and result types without looking at the body.
 bool shouldTypeCheckInEnclosingExpression(ClosureExpr *expr);
 
+/// Visit each subexpression that will be part of the constraint system
+/// of the given expression, including those in closure bodies that will be
+/// part of the constraint system/
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.h:5498]: Nit: `/` -> `.`?

Comment (DougGregor) [reply lib/Sema/ConstraintSystem.h]: Fixed, thanks!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci smoke test platform macOS

**Comment by DougGregor** [comment]
@swift-ci smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by xedin — APPROVED** [review]

**Pull request merged**